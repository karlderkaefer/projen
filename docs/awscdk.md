# AWS Cloud Projects

We support two types of projects for cloud development powered by the AWS Cloud
Development Kit (AWS CDK): **apps** and **libraries**. Apps represent complete
cloud applications while libraries vend constructs which can be consumed by
other libraries or by apps. Libraries are published to public or internal
package managers (npm, PyPI, Maven, NuGet, etc) while apps are deployed into AWS
environments.

This section describes features that are available in both cloud libraries and
applications. See [AWS CDK Construct Library](./awscdk-construct.md) and [AWS
CDK Applications](./awscdk-apps.md) for specific details about libraries and
applications.

## AWS Lambda Functions

AWS Lambda is a serverless compute platform which executes short running code
within a managed runtime environment.

To define AWS Lambda functions, create a file with a `.lambda.ts` suffix under
the source tree with AWS Lambda handler code.

For example, say we create `src/resize-image.lambda.ts` with the following
content:

```ts
export async function handler(event: any) {
  console.log('I am resizing the image now!');
}
```

Now run:

```sh
$ npx projen
```

You'll notice that a new file `src/resize-image-function.ts` has been added to
your project. This is a generated source file which exports a construct named
`ResizeImageFunction`. This construct is a subclass of
`@aws-cdk/aws-lambda.Function`, bound to your specific handler. This means that
you don't need to specify neither the `code` nor the `runtime` options when you
add it to your app:

```ts
import { ResizeImageFunction } from './resize-image-function.ts';

const handler = new ResizeImageFunction(this, 'ResizeImageFunction', {
  env: {
    FOO: '1234',
  },

  // all lambda options are supported...
});
```

Under the hood, we also added a compilation task to your project which creates a
.zip bundle for each handler. This bundle is created with
[esbuild](https://github.com/evanw/esbuild) and includes only your handler code
and all of its dependencies. This means that you can freely install and use any
dependencies in your project and use them in your handlers. You can manually
bundle your handler by executing the `bundle:HANDLER` or `bundle:watch:HANDLER`
tasks.

To customize this behavior for all functions, use `lambdaOptions` at the project
level. For example:

```ts
const { awscdk } = require('projen');

new AwsCdkConstructLibrary({
  // ...
  lambdaOptions: {
    // target node.js runtime
    runtime: awscdk.LambdaRuntime.NODEJS_14_X,

    bundlingOptions: {
      // list of node modules to exclude from the bundle
      externals: [ 'aws-sdk' ],
      sourcemap: true,
    }
  }
});
```

You can also disable auto-discovery by setting `lambdaAutoDiscover` to
`false` and then create explicitly add a `awscdk.LambdaFunction` component for
each function in your project. This will allow you to perform more
customizations as needed.

```ts
const { awscdk } = require('projen');

const p = new AwsCdkTypeScriptApp({
  lambdaAutoDiscover: false
});

new awscdk.LambdaFunction(p, {
  entrypoint: 'src/foo.lambda.ts', // .lambda.ts extension is still required
  runtime: aws_lambda.Runtime.NODEJS_12_X,
});
```

## Integration Snapshot Tests

Files in the `test/` tree with the `.integ.ts` suffix are recognized as
*integration snapshot tests*.

Each test is a simple CDK app (e.g. calls `app.synth()`) which exercises certain
construct(s) within the project. A test is considered passing if the app can be
successfully deployed.

To create/update the snapshot, developers are expected to execute the task
`integ:NAME:deploy` with AWS credentials for their personal development
environment. This task will deploy the test app to their account. Upon
successful deployment (i.e. the test passed), the snapshot will be captured and
stored under a directory called `xxx.integ.snapshot` next to the test
entrypoint. This directory should be committed to the repository.

During builds (either local or within a workflow), the task `integ:NAME:assert`
will be executed. This task synthesizes the test app and compares the output to
the captured snapshot. The build will fail if the output differs.

For each integration test, the following set of tasks are created:

|Task|Description|
|----|-----------|
|`integ:NAME:deploy`|Deploys & destroys the test app and updates the snapshot.|
|`integ:NAME:assert`|Synthesizes the test app and compares it with the snapshot (this is the task that runs during build)|
|`integ:NAME:snapshot`|Synthesizes the test app and updates the snapshot (not recommended to use because it bypasses deployment).|
|`integ:NAME:destroy`|Destroys a previously deployed test app.|

### Writing test assertions

You can write your test assertions as AWS Lambda handlers and use [CDK
Triggers](https://github.com/awslabs/cdk-triggers) to execute them as part of
the deployment.

Here is an example of a test:

```ts
import { App,Stack } from '@aws-cdk/core';
import { ConstructUnderTest } from '../src';
import { AfterCreate } from 'cdk-triggers';
import { AssertSomeStuffFunction } from './assert-some-stuff-function.ts'; // <-- generated

const app = new App();
const stack = new Stack(app, 'Test');

// this is the construct we want to test
const testee = new ConstructUnderTest(stack, 'ConstructUnderTest');

// execute a lambda handler with some assertions after all testee 
// resources are created
new AfterCreate(stack, 'RunAssertions', {
  resources: [testee],
  handler: new AssertSomeStuffFunction(stack, 'AssertSomeStuffFunction', {
    env: {
      URL: testee.url // <-- some reference to the created construct
    }
  }),
});
```

## Watch

> Only relevant for app projects

The `watch` command will use [cdk watch] in order to trigger deployments (with
opportunistic hot-swapping) when source files or asset bundles are updated.
`cdk.json` will automatically be configured to watch both source code changes
and bundles, and rebuild bundles as needed.

[cdk watch]: https://aws.amazon.com/blogs/developer/increasing-development-speed-with-cdk-watch/

To start watching, set up your environment with AWS credentials and `AWS_REGION`
pointing to your development AWS account and execute:

```sh
npx projen watch
```

This will:

* Bundle your assets (if you have any).
* Perform an initial deployment of your app into your development environment.
* Start watching for changes.

If you change a source file in your project, this change will be picked up by
`cdk watch`, assets will be re-bundled and a hotswap deployment will be
performed. For example, if you only change some AWS Lambda code, the CDK CLI
will simply update the AWS Lambda service with the location of your new code
bundle instead of going through an AWS CloudFormation deployment.

## Roadmap

* Additional bundling targets: web apps, ECS
* Local execution for AWS Lambda, ECS containers, Step Functions
* Support different provisioning engines (CloudFormation/Terraform) using Terraform L2 support
* Generate types for strong-typing AWS Lambda/ECS environment bindings.
