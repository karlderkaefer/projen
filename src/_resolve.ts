import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
