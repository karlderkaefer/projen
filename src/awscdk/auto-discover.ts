import {
  AutoDiscoverBase,
  IntegrationTestAutoDiscoverBase,
  IntegrationTestAutoDiscoverBaseOptions,
} from "../cdk";
import { Component } from "../component";
import { Project } from "../project";
import { AwsCdkDeps } from "./awscdk-deps";
import {
  IntegrationTest,
  IntegrationTestCommonOptions,
} from "./integration-test";
import { TYPESCRIPT_LAMBDA_EXT } from "./internal";
import { LambdaFunction, LambdaFunctionCommonOptions } from "./lambda-function";

/**
 * Common options for auto discovering project subcomponents.
 */
export interface AutoDiscoverCommonOptions {
  /**
   * Path to the tsconfig file to use for integration tests.
   */
  readonly tsconfigPath: string;

  /**
   * AWS CDK dependency manager.
   */
  readonly cdkDeps: AwsCdkDeps;
}

/**
 * Options for `IntegrationTestAutoDiscover`
 */
export interface IntegrationTestAutoDiscoverOptions
  extends AutoDiscoverCommonOptions,
    IntegrationTestAutoDiscoverBaseOptions {
  /**
   * Options for integration tests.
   */
  readonly integrationTestOptions?: IntegrationTestCommonOptions;
}

/**
 * Creates integration tests from entry points discovered in the test tree.
 */
export class IntegrationTestAutoDiscover extends IntegrationTestAutoDiscoverBase {
  constructor(project: Project, options: IntegrationTestAutoDiscoverOptions) {
    super(project, options);

    for (const entrypoint of this.entrypoints) {
      new IntegrationTest(this.project, {
        entrypoint,
        cdkDeps: options.cdkDeps,
        tsconfigPath: options.tsconfigPath,
        ...options.integrationTestOptions,
      });
    }
  }
}

/**
 * Options for `LambdaAutoDiscover`
 */
export interface LambdaAutoDiscoverOptions extends AutoDiscoverCommonOptions {
  /**
   * Project source tree (relative to project output directory).
   */
  readonly srcdir: string;

  /**
   * Options for auto-discovery of AWS Lambda functions.
   */
  readonly lambdaOptions?: LambdaFunctionCommonOptions;
}

/**
 * Creates lambdas from entry points discovered in the project's source tree.
 */
export class LambdaAutoDiscover extends AutoDiscoverBase {
  constructor(project: Project, options: LambdaAutoDiscoverOptions) {
    super(project, {
      projectdir: options.srcdir,
      extension: TYPESCRIPT_LAMBDA_EXT,
    });

    for (const entrypoint of this.entrypoints) {
      new LambdaFunction(this.project, {
        entrypoint,
        cdkDeps: options.cdkDeps,
        ...options.lambdaOptions,
      });
    }
  }
}

/**
 * Options for `AutoDiscover`
 */
export interface AutoDiscoverOptions
  extends LambdaAutoDiscoverOptions,
    IntegrationTestAutoDiscoverOptions {
  /**
   * Auto-discover lambda functions.
   *
   * @default true
   */
  readonly lambdaAutoDiscover?: boolean;

  /**
   * Auto-discover integration tests.
   *
   * @default true
   */
  readonly integrationTestAutoDiscover?: boolean;
}

/**
 * Discovers and creates integration tests and lambdas from code in the
 * project's source and test trees.
 */
export class AutoDiscover extends Component {
  constructor(project: Project, options: AutoDiscoverOptions) {
    super(project);

    if (options.lambdaAutoDiscover ?? true) {
      new LambdaAutoDiscover(this.project, {
        cdkDeps: options.cdkDeps,
        tsconfigPath: options.tsconfigPath,
        srcdir: options.srcdir,
        lambdaOptions: options.lambdaOptions,
      });
    }

    if (options.integrationTestAutoDiscover ?? true) {
      new IntegrationTestAutoDiscover(this.project, {
        cdkDeps: options.cdkDeps,
        testdir: options.testdir,
        tsconfigPath: options.tsconfigPath,
        integrationTestOptions: options.integrationTestOptions,
      });
    }
  }
}
