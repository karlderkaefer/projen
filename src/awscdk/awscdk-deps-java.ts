import { AwsCdkDeps, AwsCdkPackageNames } from "./awscdk-deps";

/**
 * Manages dependencies on the AWS CDK for Java projects.
 */
export class AwsCdkDepsJava extends AwsCdkDeps {
  protected packageNames(): AwsCdkPackageNames {
    return {
      coreV1: "software.amazon.awscdk/core",
      coreV2: "software.amazon.awscdk/aws-cdk-lib",
      constructs: "software.constructs/constructs",
      assertions: "software.amazon.awscdk/assertions",
    };
  }
}
