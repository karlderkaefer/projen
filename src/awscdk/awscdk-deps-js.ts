import { AwsCdkDeps, AwsCdkPackageNames } from "./awscdk-deps";

/**
 * Manages dependencies on the AWS CDK for Node.js projects.
 */
export class AwsCdkDepsJs extends AwsCdkDeps {
  protected packageNames(): AwsCdkPackageNames {
    return {
      coreV1: "@aws-cdk/core",
      coreV2: "aws-cdk-lib",
      constructs: "constructs",
      assert: "@aws-cdk/assert",
      assertions: "@aws-cdk/assertions",
    };
  }
}
