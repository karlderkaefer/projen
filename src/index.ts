export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
