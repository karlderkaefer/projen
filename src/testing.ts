import { Project } from "./project";
import { synthSnapshot } from "./util/synth";

/**
 * A Testing static class with a .synth helper for getting a snapshots of construct outputs.
 * Useful for snapshot testing with Jest.
 *
 * @example `expect(Testing.synth(someProject)).toMatchSnapshot()`
 */
export class Testing {
  /**
   * Produces a simple JS object that represents the contents of the projects with field names being file paths.
   * @param project the project to produce a snapshot for
   * @return { [filename:string]: any }
   */
  public static synth(project: Project): Record<string, any> {
    return synthSnapshot(project);
  }

  private constructor() {} // utility
}
