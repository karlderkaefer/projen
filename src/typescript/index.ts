export * from "./projenrc";
export * from "./typescript";
export * from "./typescript-typedoc";
