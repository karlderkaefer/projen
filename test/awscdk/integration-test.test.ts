import { awscdk, DependencyType } from "../../src";
import { IntegrationTest } from "../../src/awscdk";
import { AwsCdkDepsJs } from "../../src/awscdk/awscdk-deps-js";
import { Testing } from "../../src/testing";
import { TypeScriptProject } from "../../src/typescript";

describe("IntegrationTest", () => {
  // GIVEN
  const project = new awscdk.AwsCdkTypeScriptApp({
    name: "test",
    defaultReleaseBranch: "main",
    cdkVersion: "1.134.0",
  });

  // WHEN
  new awscdk.IntegrationTest(project, {
    entrypoint: "test/foo.integ.ts",
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: project.cdkDeps,
  });

  new awscdk.IntegrationTest(project, {
    entrypoint: "test/bar.integ.ts",
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: project.cdkDeps,
  });

  // THEN
  const output = Testing.synth(project);

  // we expect .npmignore to exclude the integration test's cdkout directory
  // and the various temporary directories created during execution.
  test("npmignore", () => {
    ["test/.tmp", "test/foo.integ.snapshot"].forEach((i) =>
      expect(output[".npmignore"]).toContain(i)
    );
  });

  // exclude cloud assembly manifests and assets from
  // resulting assembly (as well as nested assemblies)
  // but include cloudformation templates.
  test("gitignore", () => {
    [
      "test/foo.integ.snapshot/asset.*",
      "test/foo.integ.snapshot/**/asset.*",
      "test/foo.integ.snapshot/cdk.out",
      "test/foo.integ.snapshot/**/cdk.out",
      "test/foo.integ.snapshot/manifest.json",
      "test/foo.integ.snapshot/**/manifest.json",
      "test/foo.integ.snapshot/tree.json",
      "test/foo.integ.snapshot/**/tree.json",
      "test/.tmp",
    ].forEach((i) => expect(output[".gitignore"]).toContain(i));
  });

  test("tasks", () => {
    // list of expected tasks
    const expectedTaskNames = [
      "integ:foo:assert",
      "integ:foo:deploy",
      "integ:foo:destroy",
      "integ:foo:snapshot",
      "integ:foo:watch",
      "integ:snapshot-all",
    ];

    const actualTaskNames = Object.keys(output[".projen/tasks.json"].tasks);
    for (const t of expectedTaskNames) {
      expect(actualTaskNames).toContain(t);
      expect(output[".projen/tasks.json"].tasks[t]).toMatchSnapshot();
    }
  });
});

test("installs ts-node if needed", () => {
  const project = new TypeScriptProject({
    name: "test",
    defaultReleaseBranch: "main",
  });

  new IntegrationTest(project, {
    entrypoint: "test/foo.integ.ts",
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: new AwsCdkDepsJs(project, {
      cdkVersion: "1.0.0",
      dependencyType: DependencyType.RUNTIME,
    }),
  });

  expect(project.deps.getDependency("ts-node")).toStrictEqual({
    name: "ts-node",
    type: "build",
  });
});

test("installs aws-cdk v1 if needed", () => {
  const project = new TypeScriptProject({
    name: "test",
    defaultReleaseBranch: "main",
  });

  new IntegrationTest(project, {
    entrypoint: "test/foo.integ.ts",
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: new AwsCdkDepsJs(project, {
      cdkVersion: "1.0.0",
      dependencyType: DependencyType.RUNTIME,
    }),
  });

  expect(project.deps.getDependency("aws-cdk")).toStrictEqual({
    name: "aws-cdk",
    type: "build",
    version: "^1",
  });
});

test("installs aws-cdk v2 if needed", () => {
  const project = new TypeScriptProject({
    name: "test",
    defaultReleaseBranch: "main",
  });

  new IntegrationTest(project, {
    entrypoint: "test/foo.integ.ts",
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: new AwsCdkDepsJs(project, {
      cdkVersion: "2.8.0",
      dependencyType: DependencyType.RUNTIME,
    }),
  });

  expect(project.deps.getDependency("aws-cdk")).toStrictEqual({
    name: "aws-cdk",
    type: "build",
    version: "^2",
  });
});

test("synthesizing cdk v2 integration tests", () => {
  // GIVEN
  const project = new awscdk.AwsCdkTypeScriptApp({
    name: "test",
    defaultReleaseBranch: "main",
    cdkVersion: "2.3.1",
  });

  // WHEN
  new awscdk.IntegrationTest(project, {
    entrypoint: "test/foo.integ.ts",
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: project.cdkDeps,
  });

  // THEN
  const output = Testing.synth(project);

  expect(
    output[".projen/tasks.json"].tasks["integ:foo:deploy"]
  ).toMatchSnapshot();
  expect(
    output[".projen/tasks.json"].tasks["integ:foo:snapshot"]
  ).toMatchSnapshot();
  expect(
    output[".projen/tasks.json"].tasks["integ:foo:watch"]
  ).toMatchSnapshot();
});

test("synthesizing an integration test containing a multi-stack stage", () => {
  // GIVEN
  const project = new awscdk.AwsCdkTypeScriptApp({
    name: "test",
    defaultReleaseBranch: "main",
    cdkVersion: "2.3.1",
  });

  // WHEN
  new awscdk.IntegrationTest(project, {
    name: "my-stage",
    entrypoint: "test/my-stage.myinteg.ts",
    stacks: ["my-stage/*"],
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: project.cdkDeps,
  });

  // THEN
  const output = Testing.synth(project);

  expect(
    output[".projen/tasks.json"].tasks["integ:my-stage:deploy"]
  ).toMatchSnapshot();
  expect(
    output[".projen/tasks.json"].tasks["integ:my-stage:snapshot"]
  ).toMatchSnapshot();
  expect(
    output[".projen/tasks.json"].tasks["integ:my-stage:watch"]
  ).toMatchSnapshot();
});

test("enabling path metadata", () => {
  // GIVEN
  const project = new awscdk.AwsCdkTypeScriptApp({
    name: "test",
    defaultReleaseBranch: "main",
    cdkVersion: "2.3.1",
  });

  // WHEN
  new awscdk.IntegrationTest(project, {
    name: "my-stage",
    entrypoint: "test/my-stage.myinteg.ts",
    stacks: ["my-stage/*"],
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: project.cdkDeps,
    pathMetadata: true,
  });

  // THEN
  const output = Testing.synth(project);

  expect(
    output[".projen/tasks.json"].tasks["integ:my-stage:deploy"].steps
  ).not.toEqual(
    expect.arrayContaining([
      { exec: expect.stringContaining("--no-path-metadata") },
    ])
  );

  expect(
    output[".projen/tasks.json"].tasks["integ:my-stage:snapshot"].steps
  ).not.toEqual(
    expect.arrayContaining([
      { exec: expect.stringContaining("--no-path-metadata") },
    ])
  );

  expect(
    output[".projen/tasks.json"].tasks["integ:my-stage:watch"].steps
  ).not.toEqual(
    expect.arrayContaining([
      { exec: expect.stringContaining("--no-path-metadata") },
    ])
  );
});
