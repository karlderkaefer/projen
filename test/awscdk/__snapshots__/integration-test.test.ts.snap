// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntegrationTest tasks 1`] = `
Object {
  "description": "assert the snapshot of integration test 'foo'",
  "name": "integ:foo:assert",
  "steps": Array [
    Object {
      "exec": "[ -d \\"test/foo.integ.snapshot\\" ] || (echo \\"No snapshot available for integration test 'foo'. Run 'projen integ:foo:deploy' to capture.\\" && exit 1)",
    },
    Object {
      "exec": "cdk synth --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata --context aws-cdk:enableDiffNoFail=true --context @aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId=true --context @aws-cdk/core:enableStackNameDuplicates=true --context @aws-cdk/core:stackRelativeExports=true --context @aws-cdk/aws-ecr-assets:dockerIgnoreSupport=true --context @aws-cdk/aws-secretsmanager:parseOwnedSecretName=true --context @aws-cdk/aws-kms:defaultKeyPolicies=true --context @aws-cdk/aws-s3:grantWriteWithoutAcl=true --context @aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount=true --context @aws-cdk/aws-rds:lowercaseDbIdentifier=true --context @aws-cdk/aws-efs:defaultEncryptionAtRest=true --context @aws-cdk/aws-lambda:recognizeVersionProps=true --context @aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021=true --context @aws-cdk/core:newStyleStackSynthesis=true -o test/.tmp/foo.integ/assert.cdk.out > /dev/null",
    },
    Object {
      "exec": "diff -r -x asset.* -x cdk.out -x manifest.json -x tree.json test/foo.integ.snapshot/ test/.tmp/foo.integ/assert.cdk.out/",
    },
  ],
}
`;

exports[`IntegrationTest tasks 2`] = `
Object {
  "description": "deploy integration test 'foo' and capture snapshot",
  "name": "integ:foo:deploy",
  "steps": Array [
    Object {
      "exec": "rm -fr test/.tmp/foo.integ/deploy.cdk.out",
    },
    Object {
      "exec": "cdk deploy --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata --context aws-cdk:enableDiffNoFail=true --context @aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId=true --context @aws-cdk/core:enableStackNameDuplicates=true --context @aws-cdk/core:stackRelativeExports=true --context @aws-cdk/aws-ecr-assets:dockerIgnoreSupport=true --context @aws-cdk/aws-secretsmanager:parseOwnedSecretName=true --context @aws-cdk/aws-kms:defaultKeyPolicies=true --context @aws-cdk/aws-s3:grantWriteWithoutAcl=true --context @aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount=true --context @aws-cdk/aws-rds:lowercaseDbIdentifier=true --context @aws-cdk/aws-efs:defaultEncryptionAtRest=true --context @aws-cdk/aws-lambda:recognizeVersionProps=true --context @aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021=true --context @aws-cdk/core:newStyleStackSynthesis=true '**' --require-approval=never -o test/.tmp/foo.integ/deploy.cdk.out",
    },
    Object {
      "exec": "rm -fr test/foo.integ.snapshot",
    },
    Object {
      "exec": "mv test/.tmp/foo.integ/deploy.cdk.out test/foo.integ.snapshot",
    },
    Object {
      "spawn": "integ:foo:destroy",
    },
  ],
}
`;

exports[`IntegrationTest tasks 3`] = `
Object {
  "description": "destroy integration test 'foo'",
  "name": "integ:foo:destroy",
  "steps": Array [
    Object {
      "exec": "cdk destroy --app test/foo.integ.snapshot '**' --no-version-reporting",
    },
  ],
}
`;

exports[`IntegrationTest tasks 4`] = `
Object {
  "description": "update snapshot for integration test \\"foo\\"",
  "name": "integ:foo:snapshot",
  "steps": Array [
    Object {
      "exec": "cdk synth --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata --context aws-cdk:enableDiffNoFail=true --context @aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId=true --context @aws-cdk/core:enableStackNameDuplicates=true --context @aws-cdk/core:stackRelativeExports=true --context @aws-cdk/aws-ecr-assets:dockerIgnoreSupport=true --context @aws-cdk/aws-secretsmanager:parseOwnedSecretName=true --context @aws-cdk/aws-kms:defaultKeyPolicies=true --context @aws-cdk/aws-s3:grantWriteWithoutAcl=true --context @aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount=true --context @aws-cdk/aws-rds:lowercaseDbIdentifier=true --context @aws-cdk/aws-efs:defaultEncryptionAtRest=true --context @aws-cdk/aws-lambda:recognizeVersionProps=true --context @aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021=true --context @aws-cdk/core:newStyleStackSynthesis=true -o test/foo.integ.snapshot > /dev/null",
    },
  ],
}
`;

exports[`IntegrationTest tasks 5`] = `
Object {
  "description": "watch integration test 'foo' (without updating snapshots)",
  "name": "integ:foo:watch",
  "steps": Array [
    Object {
      "exec": "cdk watch --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata --context aws-cdk:enableDiffNoFail=true --context @aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId=true --context @aws-cdk/core:enableStackNameDuplicates=true --context @aws-cdk/core:stackRelativeExports=true --context @aws-cdk/aws-ecr-assets:dockerIgnoreSupport=true --context @aws-cdk/aws-secretsmanager:parseOwnedSecretName=true --context @aws-cdk/aws-kms:defaultKeyPolicies=true --context @aws-cdk/aws-s3:grantWriteWithoutAcl=true --context @aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount=true --context @aws-cdk/aws-rds:lowercaseDbIdentifier=true --context @aws-cdk/aws-efs:defaultEncryptionAtRest=true --context @aws-cdk/aws-lambda:recognizeVersionProps=true --context @aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021=true --context @aws-cdk/core:newStyleStackSynthesis=true '**' -o test/.tmp/foo.integ/deploy.cdk.out",
    },
  ],
}
`;

exports[`IntegrationTest tasks 6`] = `
Object {
  "description": "update snapshot for all integration tests",
  "name": "integ:snapshot-all",
  "steps": Array [
    Object {
      "spawn": "integ:foo:snapshot",
    },
    Object {
      "spawn": "integ:bar:snapshot",
    },
  ],
}
`;

exports[`synthesizing an integration test containing a multi-stack stage 1`] = `
Object {
  "description": "deploy integration test 'my-stage' and capture snapshot",
  "name": "integ:my-stage:deploy",
  "steps": Array [
    Object {
      "exec": "rm -fr test/.tmp/my-stage.integ/deploy.cdk.out",
    },
    Object {
      "exec": "cdk deploy --app \\"ts-node -P tsconfig.dev.json test/my-stage.myinteg.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata 'my-stage/*' --require-approval=never -o test/.tmp/my-stage.integ/deploy.cdk.out",
    },
    Object {
      "exec": "rm -fr test/my-stage.integ.snapshot",
    },
    Object {
      "exec": "mv test/.tmp/my-stage.integ/deploy.cdk.out test/my-stage.integ.snapshot",
    },
    Object {
      "spawn": "integ:my-stage:destroy",
    },
  ],
}
`;

exports[`synthesizing an integration test containing a multi-stack stage 2`] = `
Object {
  "description": "update snapshot for integration test \\"my-stage\\"",
  "name": "integ:my-stage:snapshot",
  "steps": Array [
    Object {
      "exec": "cdk synth --app \\"ts-node -P tsconfig.dev.json test/my-stage.myinteg.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata -o test/my-stage.integ.snapshot > /dev/null",
    },
  ],
}
`;

exports[`synthesizing an integration test containing a multi-stack stage 3`] = `
Object {
  "description": "watch integration test 'my-stage' (without updating snapshots)",
  "name": "integ:my-stage:watch",
  "steps": Array [
    Object {
      "exec": "cdk watch --app \\"ts-node -P tsconfig.dev.json test/my-stage.myinteg.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata 'my-stage/*' -o test/.tmp/my-stage.integ/deploy.cdk.out",
    },
  ],
}
`;

exports[`synthesizing cdk v2 integration tests 1`] = `
Object {
  "description": "deploy integration test 'foo' and capture snapshot",
  "name": "integ:foo:deploy",
  "steps": Array [
    Object {
      "exec": "rm -fr test/.tmp/foo.integ/deploy.cdk.out",
    },
    Object {
      "exec": "cdk deploy --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata '**' --require-approval=never -o test/.tmp/foo.integ/deploy.cdk.out",
    },
    Object {
      "exec": "rm -fr test/foo.integ.snapshot",
    },
    Object {
      "exec": "mv test/.tmp/foo.integ/deploy.cdk.out test/foo.integ.snapshot",
    },
    Object {
      "spawn": "integ:foo:destroy",
    },
  ],
}
`;

exports[`synthesizing cdk v2 integration tests 2`] = `
Object {
  "description": "update snapshot for integration test \\"foo\\"",
  "name": "integ:foo:snapshot",
  "steps": Array [
    Object {
      "exec": "cdk synth --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata -o test/foo.integ.snapshot > /dev/null",
    },
  ],
}
`;

exports[`synthesizing cdk v2 integration tests 3`] = `
Object {
  "description": "watch integration test 'foo' (without updating snapshots)",
  "name": "integ:foo:watch",
  "steps": Array [
    Object {
      "exec": "cdk watch --app \\"ts-node -P tsconfig.dev.json test/foo.integ.ts\\" --no-version-reporting --no-asset-metadata --no-path-metadata '**' -o test/.tmp/foo.integ/deploy.cdk.out",
    },
  ],
}
`;
